_posts/2015-11-23-elm-for-javascript-developers.md

```diff
@@ -0,0 +1,111 @@
+---
+layout: post
+title: Elm for JavaScript Developers
+intro: In this post we'll explore Elm, a new programming language aimed at frontend developers.
+---
+
+If you follow me on GitHub or Twitter you'll have noticed that I've been doing a lot of work with [Elm](http://elm-lang.org/) recently. Elm is a new language aimed at making it easier to build more robust, complex applications. It compiles to JavaScript but shares very little in common with the language, and its syntax will look familiar to anyone who's worked with Haskell. In the first of many posts about Elm, I'll talk through some of the major features of the language and why you should consider giving it a try. Don't be put off by its different syntax; once you get used to it you'll realiise that it's a pleasure to work with.
+
+## Immutablility and Pure Functions
+
+Every single piece of data you have in your Elm application is immutable. This means that it can never be modified, and will always be set to the value it was given when it was created. What this means in practice is that code is much easier to follow, because you know it hasn't changed. As an example, think about the below JavaScript code:
+
+```js
+var person = { name: 'Jack' };
+doSomethingWith(person);
+console.log(person);
+```
+
+Without executing that code, are you able to make any guarantees about the value of `person` once `doSomethingWith` has run? Because objects in JavaScript are mutable, you can't. `doSomethingWith` could have done anything to `person`. In a larger application where objects are frequently passed through functions bugs are often caused by this. Functions that modify the state of the world, by mutating variables given to it, or that are globally available, are functions with __side effects__. Functions like this are difficult to debug and hard to work with. 
+
+In Elm, every function is __pure__. This means two things:
+
+- Given an input X, it will always result in output Y. If you give a function the same value, it will always produce the same result.
+- The function has no side effects, and does not mutate anything or change the state of the world around it.
+
+It's of course possible to create functions like this in JavaScript, and you can make it a rule in your application that functions should be pure. Elm enforces it due to its immutable nature, and this means it's impossible for impure functions to sneak into your code base, either through code you write or through code in a 3rd party library you're using.
```

Comment:
`It's` -> `It is`

I think it's more common to split the `it` and the `is` when they precede `of course`. Just sounds a bit more natural?